dronegen/container_images.go

```diff
@@ -0,0 +1,942 @@
+	PublicEcrRegion       string = "us-east-1"
+	StagingEcrRegion      string = "us-west-2"
+
+	localRegistry string = "drone-docker-registry:5000"
+)
+
+func buildContainerImagePipelines() []pipeline {
+	// *************************************************************
+	// ****** These need to be updated on each major release. ******
+	// *************************************************************
+	latestMajorVersions := []string{"v10", "v9", "v8"}
+	branchMajorVersion := "v10"
+
+	triggers := []*TriggerInfo{
+		NewTagTrigger(branchMajorVersion),
+		NewPromoteTrigger(branchMajorVersion),
+		NewCronTrigger(latestMajorVersions),
+	}
+
+	if configureForPRTestingOnly {
+		triggers = append(triggers, NewTestTrigger(prBranch, branchMajorVersion))
+	}
+
+	pipelines := make([]pipeline, 0, len(triggers))
+	for _, trigger := range triggers {
+		pipelines = append(pipelines, trigger.buildPipelines()...)
+	}
+
+	return pipelines
+}
+
+type TriggerInfo struct {
+	Trigger           trigger
+	Name              string
+	Flags             *TriggerFlags
+	SupportedVersions []*releaseVersion
+	SetupSteps        []step
+}
+
+// This is mainly used to make passing these vars around cleaner
+type TriggerFlags struct {
+	ShouldAffectProductionImages bool
+	ShouldBuildNewImages         bool
+	UseUniqueStagingTag          bool
+}
+
+func NewTestTrigger(triggerBranch, testMajorVersion string) *TriggerInfo {
+	baseTrigger := NewCronTrigger([]string{testMajorVersion})
+	baseTrigger.Name = "Test trigger on push"
+	baseTrigger.Trigger = trigger{
+		Repo:   triggerRef{Include: []string{"gravitational/teleport"}},
+		Event:  triggerRef{Include: []string{"push"}},
+		Branch: triggerRef{Include: []string{triggerBranch}},
+	}
+
+	return baseTrigger
+}
+
+func NewTagTrigger(branchMajorVersion string) *TriggerInfo {
+	tagTrigger := triggerTag
+
+	return &TriggerInfo{
+		Trigger: tagTrigger,
+		Name:    "tag",
+		Flags: &TriggerFlags{
+			ShouldAffectProductionImages: false,
+			ShouldBuildNewImages:         true,
+			UseUniqueStagingTag:          false,
+		},
+		SupportedVersions: []*releaseVersion{
+			{
+				MajorVersion:        branchMajorVersion,
+				ShellVersion:        "$DRONE_TAG",
+				RelativeVersionName: "branch",
+			},
+		},
+	}
+}
+
+func NewPromoteTrigger(branchMajorVersion string) *TriggerInfo {
+	promoteTrigger := triggerPromote
+	promoteTrigger.Target.Include = append(promoteTrigger.Target.Include, "promote-docker")
+
+	return &TriggerInfo{
+		Trigger: promoteTrigger,
+		Name:    "promote",
+		Flags: &TriggerFlags{
+			ShouldAffectProductionImages: true,
+			ShouldBuildNewImages:         false,
+			UseUniqueStagingTag:          false,
+		},
+		SupportedVersions: []*releaseVersion{
+			{
+				MajorVersion:        branchMajorVersion,
+				ShellVersion:        "$DRONE_TAG",
+				RelativeVersionName: "branch",
+			},
+		},
+		SetupSteps: verifyValidPromoteRunSteps(),
+	}
+}
+
+func NewCronTrigger(latestMajorVersions []string) *TriggerInfo {
```

Comment:
We can't have a separate drone.yaml for cron but can we at leat factor out all cron related stuff in a separate .go file. I left a number of suggestion around here for better code org.